src/workerd/api/streams/encoding.c++

```diff
@@ -4,35 +4,136 @@
-                      [](jsg::Lock& js, auto chunk, auto controller) {
+      Transformer{
+        .transform = jsg::Function<Transformer::TransformAlgorithm>(
+            [pending = wrapper->addWrappedRef()](jsg::Lock& js, v8::Local<v8::Value> chunk,
+                jsg::Ref<TransformStreamDefaultController> controller) mutable {
     auto str = jsg::check(chunk->ToString(js.v8Context()));
-    auto utf8Length = str->Utf8LengthV2(js.v8Isolate);
+    size_t length = str->Length();
+
+    // Early exit: empty chunk with no pending surrogate produces no output
+    if (length == 0 && *pending == kj::none) return js.resolvedPromise();
+
+    // Allocate buffer: reserve slot 0 for pending surrogate if we have one
+    size_t prefix = (*pending != kj::none) ? 1 : 0;
+    auto buf = kj::heapArray<char16_t>(prefix + length);
+    str->WriteV2(js.v8Isolate, 0, length, reinterpret_cast<uint16_t*>(buf.begin() + prefix));
+
+    // Handle pending high surrogate from previous chunk
+    KJ_IF_SOME(lead, *pending) {
+      KJ_DASSERT(high_surrogate(lead), "pending must be a high surrogate");
+      // Empty chunk: keep pending surrogate for next chunk
+      if (length == 0) return js.resolvedPromise();
+      *pending = kj::none;
+      // If chunk starts with matching low surrogate, complete the pair; otherwise emit U+FFFD
+      buf[0] = low_surrogate(buf[prefix]) ? lead : REPLACEMENT_CHAR;
+    }
+
+    size_t end = prefix + length;
+    KJ_DASSERT(end <= buf.size() && end > 0);
 
-    // Don't emit empty chunks
-    if (utf8Length == 0) {
-      return js.resolvedPromise();
+    // If chunk ends with high surrogate, save it for next chunk
+    if (end > 0 && high_surrogate(buf[end - 1])) {
+      *pending = buf[--end];
     }
 
-    v8::Local<v8::ArrayBuffer> buffer;
-    JSG_REQUIRE(v8::ArrayBuffer::MaybeNew(js.v8Isolate, utf8Length).ToLocal(&buffer), RangeError,
-        "Cannot allocate space for TextEncoder.encode");
+    // Nothing to encode after handling surrogates
+    if (end == 0) return js.resolvedPromise();
+
+    // Sanitize any remaining unpaired surrogates in the middle of the buffer
+    auto slice = buf.first(end);
+    KJ_DASSERT(slice.size() > 0);
+    simdutf::to_well_formed_utf16(slice.begin(), slice.size(), slice.begin());
+
+    // Convert UTF-16 to UTF-8
+    auto utf8Length = simdutf::utf8_length_from_utf16(slice.begin(), slice.size());
+    KJ_DASSERT(utf8Length > 0);
 
-    auto bytes = jsg::asBytes(buffer).releaseAsChars();
-    [[maybe_unused]] auto written = str->WriteUtf8V2(
-        js.v8Isolate, bytes.begin(), bytes.size(), v8::String::WriteFlags::kReplaceInvalidUtf8);
+    auto output = jsg::BackingStore::alloc<v8::Uint8Array>(js, utf8Length);
+    auto dest = output.asArrayPtr().asChars();
+    [[maybe_unused]] auto written =
+        simdutf::convert_utf16_to_utf8(slice.begin(), slice.size(), dest.begin());
+    KJ_DASSERT(written == utf8Length, "simdutf should write exactly utf8Length bytes");
```

Comment:
Instead of always doing three steps:
```
  to_well_formed_utf16(...);
  utf8_length_from_utf16(...);
  convert_utf16_to_utf8(...);
```
you can do this in two steps in the common case:
```
  auto result = utf8_length_from_utf16_with_replacement(...);
  if (result.error == SURROGATE) to_well_formed_utf16(...);
  convert_utf16_to_utf8(...);
```